### DISCUSSION TITLE
How does the troika 3d text cliprect work?

### DISCUSSION THREAD

**Discussion by dsimonow** [protectwise/troika#207] [Q&A] [answered]
Hello everyone. 
I am currently using troika 3d text but in r3f.
It is mostly a wrapper so the cliprect behaviour shouldn't be affected but r3f could have weird interactions with it.
I am trying to replicate the scrollbarfascade functionality.

My issue is, I don't understand exactly how to manipulate cliprect. 
In the description it sounds like a rectangle that is set and everything outside will not be rendered.
But it seems as if they are percentage based and therefor the values are different for each length of a text.

Gonna attach a few images. 
They are the same object that gets different text.
With clipRect={[-10, 3, 10, 28]}
In all of them the clipping gets applied.
The text is actually as long as the white boundingbox around the text.

<img src="https://user-images.githubusercontent.com/9114996/168105689-f9a49dc8-dc00-471c-a639-c3bece37242a.png" width="200">
<img src="https://user-images.githubusercontent.com/9114996/168105692-7a38ca77-35dd-409d-9552-5b261be21a7f.png" width="200">
<img src="https://user-images.githubusercontent.com/9114996/168105696-3edcd977-9f95-46af-825c-dabf5064b6c8.png" width="200">

Is that behaviour normal and how do you solve this or is this something r3f causes?
Are the numbers percentages or how should I read them?

Thank you!

**Comment by lojjic** [comment, marked as answer]
Hmm, no the values should be absolute units in the same coordinate system as the text itself, not percentages. I can't explain the behavior in your screenshots off the top of my head, would you be able to put together a codesandbox or similar I could look at in action?

**Reply by lojjic** [reply]
Oh, wait, I think that behavior makes sense if your anchorX/anchorY are at the center of the text block (I think that may be drei's default?) The clipRect values are in the coordinate system of the text, meaning they're relative to the `[0, 0]` anchor position. So with your clipRect with min-y of 3, if your anchorY is in the center, that clips it 3 units above the center.

**Reply by dsimonow** [reply]
this is the answer. Thank you so much! 

I have another question about this.
I got it working as I wanted with anchorx/y at 0.
With clipRect={[0,-12,20,0]}.
Spanning from left to right with minX=0 and maxX=20
and for Y from the top maxY=0 down to minY=-12. 

as Example this is: clipRect={[0,-12,20,-5]}
with minY showing the first 12(pixels?) and maxY clipping the top 5.
This is the correct way how it works? 
Since 0,0 is basically top left. So anything Y becomes a negative Value?
![troikaflexsolv](https://user-images.githubusercontent.com/9114996/168147127-93bb7477-8977-4f78-b9b5-70855576672d.png)

**Reply by lojjic** [reply]
That sounds correct to me. Positive Y is up in ThreeJS's coordinate system.
